sqlmesh/core/plan/builder.py

```diff
@@ -322,56 +328,64 @@ def is_restateable_snapshot(snapshot: Snapshot) -> bool:
                         "Run the restatement against the production environment instead to restate this model."
                     )
                     continue
                 elif (not self._is_dev or not snapshot.is_paused) and snapshot.disable_restatement:
                     self._console.log_warning(
-                        f"Cannot restate model '{model_fqn}'. "
+                        f"Cannot restate model '{snapshot.name}'. "
                         "Restatement is disabled for this model to prevent possible data loss."
                         "If you want to restate this model, change the model's `disable_restatement` setting to `false`."
                     )
                     continue
                 elif snapshot.is_symbolic or snapshot.is_seed:
-                    logger.info("Skipping restatement for model '%s'", model_fqn)
+                    logger.info("Skipping restatement for model '%s'", snapshot.name)
                     continue
 
-            restatements[snapshot.snapshot_id] = dummy_interval
-            for downstream_s_id in dag.downstream(snapshot.snapshot_id):
-                if is_restateable_snapshot(self._context_diff.snapshots[downstream_s_id]):
-                    restatements[downstream_s_id] = dummy_interval
-
-        # Get restatement intervals for all restated snapshots and make sure that if an incremental snapshot expands it's
-        # restatement range that it's downstream dependencies all expand their restatement ranges as well.
-        for s_id in dag:
-            if s_id not in restatements:
-                continue
-            snapshot = self._context_diff.snapshots[s_id]
-            interval = snapshot.get_removal_interval(
-                self._start or earliest_interval_start,
-                self._end or now(),
+            removal_interval = snapshot.get_removal_interval(
+                start,
+                end,
                 self._execution_time,
                 strict=False,
                 is_preview=is_preview,
             )
+
             # Since we are traversing the graph in topological order and the largest interval range is pushed down
             # the graph we just have to check our immediate parents in the graph and not the whole upstream graph.
-            snapshot_dependencies = snapshot.parents
-            possible_intervals = [
-                restatements.get(s, dummy_interval)
-                for s in snapshot_dependencies
-                if self._context_diff.snapshots[s].is_incremental
-            ] + [interval]
+            restating_parents = [
+                self._context_diff.snapshots[s] for s in snapshot.parents if s in restatements
+            ]
+
+            if not restating_parents and snapshot.name not in restate_models:
+                continue
+            if not removal_interval:
+                self._console.log_error(
```

Comment:
If this is an error should we raise? Otherwise should we `log_warning` instead?

Reply:
should be a warning